### PR TITLE
boards: Added standalone ATmega328p

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

The PR adds the `atmega328p` "board", which allows to use the ATmega328P without any external circuitry except a stable power supply. An ICSP programmer is required to flash the board and an TTL adapter to get the serial terminal. This "board" will be much less convenient to use than an Arduino UNO or an Arduino Nano (which contain this exact MCU), but it can be very convenient as a low-power companion MCU. E.g. the ATmega328p can be used to read out a bunch of slow sensors while the main (and more power demanding) MCU is in lowest power mode. The main MCU could periodically wake up to collect and process the accumulated data from the ATmegea328p.

ATmega328p DIP package on a breadboard

### Testing procedure

- Check the provided documentation
- Using an ICSP try if `make BOARD=atmega328p flash` for `examples/hello-world` works. Refer to the doc of this PR when not using `usbtiny` as programmer
- Using an TTL adapter check if you get the output. (Keep in mind that this is not an Arduino that conveniently reboots upon `make term`...)

As this shares most code with the Arduino UNO, no huge troubles are expected.

### Issues/PRs references

None

### CODE REVIEW

**Comment by miri64** [comment]
Not really my domain ^^"

**Comment by fjmolinas** [comment]
@fjmolinas self remainder, I might have one of these laying around the office.

**Comment by fjmolinas** [comment]
@maribu Did you try testing using an arduino-uno to program? I have been trying this but fail to sync.

**Comment by maribu** [comment]
I'm not sure what you mean by "fail to sync". Did you try to flash it using the bootloader? That will not work, as the ATmega328p does not come with a bootloader by default. This PR defaults to the `usbtiny` programmer, of which some incredible cheap clones are available.

Also: Please not that the Arduino UNO has an external 16MHz crystal available and the fuses are configured to use it. This PR adds the ATmega328P as "standalone" chip without crystal by using the internal 8MHz PLL as clock source (default fuse settings when it comes out of the factory).

**Comment by maribu** [comment]
I fixed the issue with the incorrect include guard detected by CI. And I also added the ability to set the default value for `PROGRAMMER` via `PROGRAMMER_BOARD_ATMEGA328P`. That way one can run e.g. `export PROGRAMMER_BOARD_ATMEGA328P=dragon_isp` to use the Dragon ISP as default programmer of this board, without changing the behavior of other boards. (Setting `PROGRAMMER` will still take preference of this variable for consistent behavior.)

**Comment by fjmolinas** [comment]
> I'm not sure what you mean by "fail to sync". Did you try to flash it using the bootloader? That will not work, as the ATmega328p does not come with a bootloader by default. This PR defaults to the usbtiny programmer, of which some incredible cheap clones are available.

I had an arduino board form which I removed the `atmega328p`.  Since it was in an arduino board the bootloader was already loaded. I then tried flashing using an arduino (I was following this [guide](https://www.arduino.cc/en/Tutorial/ArduinoToBreadboard)). Anyway since I don't have  a `usbtiny` I won't be able to test.

Could you provide test output for the board? That would be enough for me If I can't get my hands on a programmer.

**Comment by smlng** [comment]
I tested this with an Arduino-Uno and the following command: 

```
CFLAGS="-DCLOCK_CORECLOCK=16000000UL" BAUD=9600 FFLAGS_EXTRA="-b 115200" PROGRAMMER=arduino PORT=/dev/tty.usbmodem143101 BUILD_IN_DOCKER=1 BOARD=atmega328p make -C tests/thread_flood/ flash term
```

simply inducing the flasher config via env variables and most importantly set the core clock to 16MHz otherwise shell output is scrambled. However, that works. So take this as semi-tested ACK 😉

**Comment by smlng** [comment]
could also use `ATMEGA328P_CLOCK` to set the clock (nice handling @maribu)

File: boards/atmega328p/doc.txt [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+Connect a TTL adapter with pins 2/RXD and 3/TXD an run
+
+    make BOARD=atmega328p term
+
+Please note that the supply voltage should be compatible with the logic level of
+the TTL adapter. Usually everything between 3.3 V and 5 V should work.
+
+## Caution
+Don't expect having a working network stack due to very limited resources.
```
Comment (smlng) [inline boards/atmega328p/doc.txt]: you should be a smiley here

**Comment by maribu** [comment]
@fjmolinas: Did you try changing the fuse settings while the chip was still in the arduino board? If you set the clock source to internal, it should work as standalone afterwards. If you want to still use the bootloader (by setting PROGRAMMER), I think that will work when dividing the baud rate used for flashing by two. (The shell baud rate should work untouched.)

**Comment by smlng** [comment]
please squash

**Comment by fjmolinas** [comment]
> this looks good to me, couldn't test on a bare atmega328p but as commented before tested on Arduino-uno with this board config and some simple redirects for the flasher.

I'm gonna run all tests the way @smlng configured it.

File: boards/atmega328p/doc.txt (hunk 1 above)
Comment (maribu) [reply boards/atmega328p/doc.txt]: Done

**Comment by maribu** [comment]
I added the missing smiley and squashed right away. (Click on the "force-pushed" link to see the added smiley.)

**Comment by smlng** [comment]
CI not happy, you need to update a lot of blacklist and insufficient memory for tests - basically every test not suitable for Arduino-uno will not run with this "board" either.

**Comment by fjmolinas** [comment]
I had the following failign tests:

```
python dist/tools/compile_and_test_for_board/compile_and_test_for_board.py --jobs 2 . arduino-uno --with-test-only

Failures during test:
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/libfixmath](tests/libfixmath/test.failed)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed)
- [tests/pkg_lora-serialization](tests/pkg_lora-serialization/test.failed)
- [tests/trickle](tests/trickle/test.failed)
```

**Comment by smlng** [comment]
@fjmolinas I would assume that these also fail on master with board set to `Arduino-uno`, and hence we shouldn't care so much for this PR but rather investigate and fix separately.

**Comment by smlng** [comment]
here is what I found for the failed tests:

- `bitarithm_timings`: just seems to be f**king slow, hence test fails with timeout. And/or timer problem as with trickle (see below).
- `libfixmath`: seems to work when looking at the shell, but might be too slow for auto test, too.
- `pkg_jsmn`: print problem likely related to avr libc or similar
- `pkg_lora-serialization`: likely a problem with the parameter size and values, i.e. too large for 16Bit arithmetic on atmega. Or missing size suffixes (e.g. `LU`) on macros.
- `trickle`: looks like a timer problem, as it hangs after trickle reset

all not directly related to this PR, so I say: don't care.

**Comment by fjmolinas** [comment]
> here is what I found for the failed tests:
> 
> bitarithm_timings: just seems to be f**king slow, hence test fails with timeout. And/or timer problem as with trickle (see below).
> libfixmath: seems to work when looking at the shell, but might be too slow for auto test, too.
> pkg_jsm�n: print problem likely related to avr libc or similar
> pkg_lora-serialization: likely a problem with the parameter size and values, i.e. too large for 16Bit arithmetic on atmega. Or missing size suffixes (e.g. LU) on macros.
> trickle: looks like a timer problem, as it hangs after trickle reset
> all not directly related to this PR, so I say: don't care.

I had launched the tests with the wrong board, here are the correct results:

```
CFLAGS="-DCLOCK_CORECLOCK=16000000UL" BAUD=9600 FFLAGS_EXTRA="-b 115200" PROGRAMMER=arduino PORT=/dev/ttyACM1 BUILD_IN_DOCKER=1 python dist/tools/compile_and_test_for_board/compile_and_test_for_board.py --jobs 2 . atmega328p --with-test-only

```

```
Failures during compilation:
- [tests/bloom_bytes](tests/bloom_bytes/compilation.failed)
- [tests/cond_order](tests/cond_order/compilation.failed)
- [tests/events](tests/events/compilation.failed)
- [tests/evtimer_msg](tests/evtimer_msg/compilation.failed)
- [tests/evtimer_underflow](tests/evtimer_underflow/compilation.failed)
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/compilation.failed)
- [tests/gnrc_ipv6_fwd_w_sub](tests/gnrc_ipv6_fwd_w_sub/compilation.failed)
- [tests/gnrc_ipv6_nib](tests/gnrc_ipv6_nib/compilation.failed)
- [tests/gnrc_ipv6_nib_6ln](tests/gnrc_ipv6_nib_6ln/compilation.failed)
- [tests/gnrc_ndp](tests/gnrc_ndp/compilation.failed)
- [tests/gnrc_netif](tests/gnrc_netif/compilation.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/compilation.failed)
- [tests/gnrc_sixlowpan](tests/gnrc_sixlowpan/compilation.failed)
- [tests/gnrc_sixlowpan_frag](tests/gnrc_sixlowpan_frag/compilation.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/compilation.failed)
- [tests/gnrc_sock_ip](tests/gnrc_sock_ip/compilation.failed)
- [tests/gnrc_sock_udp](tests/gnrc_sock_udp/compilation.failed)
- [tests/libfixmath_unittests](tests/libfixmath_unittests/compilation.failed)
- [tests/lwip_sock_ip](tests/lwip_sock_ip/compilation.failed)
- [tests/lwip_sock_tcp](tests/lwip_sock_tcp/compilation.failed)
- [tests/lwip_sock_udp](tests/lwip_sock_udp/compilation.failed)
- [tests/msg_send_receive](tests/msg_send_receive/compilation.failed)
- [tests/mutex_order](tests/mutex_order/compilation.failed)
- [tests/netdev_test](tests/netdev_test/compilation.failed)
- [tests/nhdp](tests/nhdp/compilation.failed)
- [tests/periph_eeprom](tests/periph_eeprom/compilation.failed)
- [tests/periph_gpio](tests/periph_gpio/compilation.failed)
- [tests/pipe](tests/pipe/compilation.failed)
- [tests/pkg_c25519](tests/pkg_c25519/compilation.failed)
- [tests/pkg_cifra](tests/pkg_cifra/compilation.failed)
- [tests/pkg_cn-cbor](tests/pkg_cn-cbor/compilation.failed)
- [tests/pkg_hacl](tests/pkg_hacl/compilation.failed)
- [tests/pkg_heatshrink](tests/pkg_heatshrink/compilation.failed)
- [tests/pkg_libb2](tests/pkg_libb2/compilation.failed)
- [tests/pkg_libcoap](tests/pkg_libcoap/compilation.failed)
- [tests/pkg_libcose](tests/pkg_libcose/compilation.failed)
- [tests/pkg_littlefs](tests/pkg_littlefs/compilation.failed)
- [tests/pkg_micro-ecc](tests/pkg_micro-ecc/compilation.failed)
- [tests/pkg_monocypher](tests/pkg_monocypher/compilation.failed)
- [tests/pkg_qdsa](tests/pkg_qdsa/compilation.failed)
- [tests/pkg_relic](tests/pkg_relic/compilation.failed)
- [tests/pkg_spiffs](tests/pkg_spiffs/compilation.failed)
- [tests/pkg_tiny-asn1](tests/pkg_tiny-asn1/compilation.failed)
- [tests/pkg_tinycbor](tests/pkg_tinycbor/compilation.failed)
- [tests/pkg_tweetnacl](tests/pkg_tweetnacl/compilation.failed)
- [tests/pkg_u8g2](tests/pkg_u8g2/compilation.failed)
- [tests/pkg_ubasic](tests/pkg_ubasic/compilation.failed)
- [tests/pkg_ucglib](tests/pkg_ucglib/compilation.failed)
- [tests/posix_semaphore](tests/posix_semaphore/compilation.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/compilation.failed)
- [tests/pthread](tests/pthread/compilation.failed)
- [tests/pthread_barrier](tests/pthread_barrier/compilation.failed)
- [tests/pthread_cleanup](tests/pthread_cleanup/compilation.failed)
- [tests/pthread_condition_variable](tests/pthread_condition_variable/compilation.failed)
- [tests/pthread_cooperation](tests/pthread_cooperation/compilation.failed)
- [tests/pthread_rwlock](tests/pthread_rwlock/compilation.failed)
- [tests/pthread_tls](tests/pthread_tls/compilation.failed)
- [tests/rmutex](tests/rmutex/compilation.failed)
- [tests/rng](tests/rng/compilation.failed)
- [tests/shell](tests/shell/compilation.failed)
- [tests/ssp](tests/ssp/compilation.failed)
- [tests/struct_tm_utility](tests/struct_tm_utility/compilation.failed)
- [tests/sys_crypto](tests/sys_crypto/compilation.failed)
- [tests/thread_cooperation](tests/thread_cooperation/compilation.failed)
- [tests/thread_exit](tests/thread_exit/compilation.failed)
- [tests/thread_flags](tests/thread_flags/compilation.failed)
- [tests/thread_msg](tests/thread_msg/compilation.failed)
- [tests/thread_msg_block_race](tests/thread_msg_block_race/compilation.failed)
- [tests/thread_msg_seq](tests/thread_msg_seq/compilation.failed)
- [tests/unittests](tests/unittests/compilation.failed)
- [tests/xtimer_hang](tests/xtimer_hang/compilation.failed)
- [tests/xtimer_msg](tests/xtimer_msg/compilation.failed)
- [tests/xtimer_periodic_wakeup](tests/xtimer_periodic_wakeup/compilation.failed)

Failures during test:
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/driver_grove_ledbar](tests/driver_grove_ledbar/test.failed)
- [tests/driver_my9221](tests/driver_my9221/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/isr_yield_higher](tests/isr_yield_higher/test.failed)
- [tests/l2util](tests/l2util/test.failed)
- [tests/libfixmath](tests/libfixmath/test.failed)
- [tests/log_printfnoformat](tests/log_printfnoformat/test.failed)
- [tests/periph_timer](tests/periph_timer/test.failed)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed)
- [tests/pkg_libhydrogen](tests/pkg_libhydrogen/test.failed)
- [tests/pkg_lora-serialization](tests/pkg_lora-serialization/test.failed)
- [tests/pkg_umorse](tests/pkg_umorse/test.failed)
- [tests/riotboot_hdr](tests/riotboot_hdr/test.failed)
- [tests/test_tools](tests/test_tools/test.failed)
```

All the failed compilations where already shown by murdock. Note that I only ran for applications with tests, so there are more no-compiling applications but those will be taken care of by murdock.

The failing tests have corrupted outputs, not sure why yet.

PS: sorry I miss-deleted my other test output.

**Comment by fjmolinas** [comment]
> bitarithm_timings: just seems to be f**king slow, hence test fails with timeout. And/or timer problem as with trickle (see below).
> libfixmath: seems to work when looking at the shell, but might be too slow for auto test, too.
> pkg_jsm�n: print problem likely related to avr libc or similar
> pkg_lora-serialization: likely a problem with the parameter size and values, i.e. too large for 16Bit arithmetic on atmega. Or missing size suffixes (e.g. LU) on macros.
> trickle: looks like a timer problem, as it hangs after trickle reset

This still applies to the new test results.

**Comment by smlng** [comment]
the tests with corrupted output somehow do not pick up the overwrite of `CLOCK_CORECLOCK` if you set that in `periph_conf.h` output is good. This seems to be a problem with `CFLAGS`, not with this PR. Again, to me this is non blocking - most tests will also fail on master someway or the other.

@maribu: you could at the board in `tests/periph_timer/Makefile` to use the 250kHz timer like arduino-uno, bc now it fails as 1MHz is not supported.

**Comment by fjmolinas** [comment]
Modifying `CLOCK_CORECLOCK` in `periph_conf.h` I have this remaining failures:

```
Failures during test:
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed) (timeouts early)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed) (RAM issue, add to BOARD_INSUFFICIENT_MEMORY)
- [tests/isr_yield_higher](tests/isr_yield_higher/test.failed) (The test script is the issue)
- [tests/libfixmath](tests/libfixmath/test.failed)   (timeouts early)
- [tests/periph_timer](tests/periph_timer/test.failed) (fixed changing TIMER_SPEED)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed) (Fails in master for arduino-uno)
- [tests/pkg_libhydrogen](tests/pkg_libhydrogen/test.failed) (need to be blacklisted)
- [tests/pkg_lora-serialization](tests/pkg_lora-serialization/test.failed) (Fails in master for arduino-uno)
```

For me as soon as the needed tests are blacklisted this is good to go.

**Comment by smlng** [comment]
@maribu please check Murdock output and add boards to insufficient memory or blacklist accordingly to make CI happy.

**Comment by maribu** [comment]
Sorry for stalling, we're super busy with FGSN right now. I'll add the missing insufficient memory and blacklist entries after FGSN and my vacation. Adding them for the Arduino-Nano and the Arduino-Leonardo was 1hr+ of stupid and exhausting work, and this will not be much different here...

**Comment by maribu** [comment]
OK, finally adding the `BOARD_INSUFFICIENT_MEMORY` entries. First, I'll rebase on master, as otherwise all the touched `Makefile`s will show merge conflicts

**Comment by maribu** [comment]
So, that was the rebase on top of master, no code in the PR was changed. Now I'll add the `BOARD_INSUFFICIENT_MEMORY` entries

**Comment by maribu** [comment]
I changed my mind. I want something like https://github.com/RIOT-OS/RIOT/pull/12406 first. I believe than I will be able to extend the lists using `find` and `sed` without fearing to mess anything else up :tada:

(The contents of proposed `Makefile.ci` are simple and very similar/predictable, which greatly reduces the chances of a regex doing something unexpected.)

**Comment by maribu** [comment]
I have a good feeling that (except for static tests due to squashing required) Murdock will have nothing to complain. May I squash?

**Comment by fjmolinas** [comment]
> I have a good feeling that (except for static tests due to squashing required) Murdock will have nothing to complain. May I squash?

Please do :)

**Comment by maribu** [comment]
I fixed this and squashed right away:

```
Running './dist/tools/buildsystem_sanity_check/check.sh' x

Command output:

	Invalid build system patterns found by ./dist/tools/buildsystem_sanity_check/check.sh:

	Makefiles files from cpu must not be included by the board anymore

		boards/atmega328p/Makefile.features:include $(RIOTCPU)/atmega328p/Makefile.features
```

**Comment by fjmolinas** [comment]
GO!

**Pull request merged**